examples/token/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+//! - State mutations.
+//! - Method dispatch.
+//! - Deserialization of parameters.
+//! - Serialization of return data.
+//!
+//! ## Addressing
+//!
+//! For simplicity, this actor uses ActorIDs in the balances and allowances
+//! HAMT (map) keys. This poses two problems:
+//!
+//! 1. A sender cannot send tokens to an inexistent account actor.
+//! 2. This actor is not reorg-safe (ActorIDs can change in the chain gets
+//!    reorg-ed).
+//!
+//! We _could_ do sophisticated things here to solve for these problems, by
+//! associating balances to class-{1,2,3} addresses only, and rejecting calls
+//! with class-0 (ID addresses).
+//!
+//! However, all of this will change when we introduce [class-4 addresses](https://github.com/filecoin-project/fvm-specs/blob/main/04-evm-mapping.md#proposed-solution-universal-stable-addresses)
+//! (reorg stable universal addresses), so it's not worth the complexity now,
+//! as this actor only exists for illustration purposes.
+//!
+//! ## NOT MEANT FOR PRODUCTION USAGE
+//!
+//! This actor should NEVER be deployed on the Filecoin network. It is purely
+//! meant for illustration and test purposes.
+//!
+
+use cid::multihash::{Code, MultihashDigest};
+use cid::Cid;
+use fvm_sdk as sdk;
+use fvm_sdk::blockstore::Blockstore;
+use fvm_sdk::message::{params_cbor, NO_DATA_BLOCK_ID};
+use fvm_shared::address::Address;
+use fvm_shared::bigint::bigint_ser::{BigIntDe, BigIntSer};
+use fvm_shared::bigint::{bigint_ser, Zero};
+use fvm_shared::blockstore::CborStore;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::tuple::*;
+use fvm_shared::encoding::{Cbor, RawBytes, DAG_CBOR};
+use fvm_shared::ActorID;
+use ipld_hamt::Hamt;
+
+/// A macro to abort execution by signalling a non-zero exit code.
+macro_rules! abort {
+    ($code:ident, $msg:literal $(, $ex:expr)*) => {
+        fvm_sdk::vm::abort(
+            fvm_shared::error::ExitCode::$code as u32,
+            Some(format!($msg, $($ex,)*).as_str()),
+        )
+    };
+}
+
+/// A macro to CBOR-encode a return value.
+macro_rules! encode {
+    ($ex:expr) => {
+        match RawBytes::serialize($ex) {
+            Ok(ret) => ret,
+            Err(err) => {
+                abort!(ErrSerialization, "failed to encode return value: {:?}", err)
+            }
+        }
+    };
+}
+
+/// The state object of the token actor.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct State {
+    pub name: String,
+    pub symbol: String,
+    #[serde(with = "bigint_ser")]
+    pub max_supply: TokenAmount,
+    // map[ActorID] => TokenAmount
+    pub balances: Cid,
+    // TODO should never use ActorID as allowees, since it's an unstable address
+    //  make sure to check for this
+    //  map[ActorID] => map[Address]TokenAmount
+    // pub allowances: Cid,
+}
+
+/// The actor's WASM entrypoint. It takes the ID of the parameters block,
+/// and returns the ID of the return value block.
+#[no_mangle]
+pub fn invoke(params_id: u32) -> u32 {
```

Comment:
We can have a lot of fun with the sugar:

```rust
#[derive(Serialize_tuple, Deserialize_tuple)]
type Actor {
    name: String,
    value: u64,
}

#[fvm_actor]
impl Actor {
    // Implicitly reads the state.
    #[fvm_export = 2]
    fn name(&self) -> &str { // takes nothing, returns a string
        &self.name
    }

    // Implicitly reads the state and writes it back
    #[fvm_export = 3]
    fn add(&mut self, value: u32) -> u32 { // takes a u32, returns a u32
        self.value += value
        self.value
    }

    // Explicitly transform the state. We could use Deref... but that sounds dangerous.
    #[fvm_export = 4]
    fn sub(state: &mut Ctx<Self>, value: u32) -> u32 { // takes a u32, returns a u32
        state.write(|st| {
            st.value -= value;
        });
        state.read(|st| st.value)
    }
}
```

We'll run into questions like:

1. Do we want some kind of FVM context object? For sending, syscalls, etc.
2. What about multiple parameters? Should we always encode parameters as a tuple? (I guess we already do).

Reply:
this is EXTREMELY groovy....... i love the idea of pure actors that just do a state monad....... i love that soooooooo much..................

Reply:
Yep, there's tons of exploration to do.